update loadbalancers service to use Azure SDK v2

nawazkh opened the pull request:
<!-- If this is your first PR, welcome! Please make sure you read the [contributing guidelines](https://github.com/kubernetes-sigs/cluster-api-provider-azure/blob/main/CONTRIBUTING.md). -->

 <!-- Please label this pull request according to what type of issue you are addressing (see ../CONTRIBUTING.md) -->
**What type of PR is this?**
/kind feature
<!--
Add one of the following kinds:

/kind bug
/kind api-change
/kind cleanup
/kind deprecation
/kind design
/kind documentation
/kind failing-test
/kind flake
-->

**What this PR does / why we need it**:
- This PR updates load balancers service to sdkv2. 

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #3917 

**Special notes for your reviewer**:
<!-- Refer to https://github.com/kubernetes-sigs/cluster-api-provider-azure/blob/main/docs/book/src/developers/releasing.md#release-support for more information about which changes are eligible for backport -->

- [ ] cherry-pick candidate

**TODOs**:
<!-- Put an "X" character inside the brackets of each completed task. Some may be optional, depending on the PR. -->

- [ ] squashed commits
- [ ] includes documentation
- [ ] adds unit tests

**Release note**:
<!--  Write your release note:
1. Enter your extended release note in the below block. If the PR requires additional action from users switching to the new release, include the string "action required".
2. If no release note is required, just write "NONE".
-->
```release-note
Update loadbalancers service to use Azure SDK v2
```

nawazkh commented:
I will squash it once the e2e go green :)

mboersma commented on azure/converters/loadbalancers.go:
```diff
@@ -17,14 +17,14 @@ limitations under the License.
 
-// SKUtoSDK converts infrav1.SKU into a network.LoadBalancerSkuName.
-func SKUtoSDK(src infrav1.SKU) network.LoadBalancerSkuName {
+// SKUtoSDK converts infrav1.SKU into a armnetwork.LoadBalancerSkuName.
```
```suggestion
// SKUtoSDK converts infrav1.SKU into an armnetwork.LoadBalancerSkuName.
```

nawazkh replied:
Updated

mboersma commented on azure/services/loadbalancers/client.go:
```diff
@@ -18,158 +18,127 @@ package loadbalancers
+		return nil, err
+	}
+
+	return resp, nil
```
```suggestion
	return resp.LoadBalancer, nil
```

This is what needs to be returned, but unfortunately the signature is just `interface{}` so we can't enforce the return value.

nawazkh replied:
Updated

mboersma commented on azure/services/loadbalancers/client.go:
```diff
@@ -18,158 +18,127 @@ package loadbalancers
-		req.Header.Add("If-Match", etag)
+	factory, err := armnetwork.NewClientFactory(ac.auth.SubscriptionID(), ac.auth.Token(), clientOpts)
+	if err != nil {
+		return nil, nil, errors.Wrap(err, "failed to create loadbalancer client factory")
```
```suggestion
		return nil, nil, errors.Wrap(err, "failed to create amrnetwork client factory")
```

nawazkh replied:
Updated

nawazkh commented:
Expected complaints from apidiff
```
sigs.k8s.io/cluster-api-provider-azure/azure/services/loadbalancers
Incompatible changes:
  - New: changed from func(LBScope) *Service to func(LBScope) (*Service, error)

sigs.k8s.io/cluster-api-provider-azure/azure/converters
Incompatible changes:
  - SKUtoSDK: changed from func(sigs.k8s.io/cluster-api-provider-azure/api/v1beta1.SKU) github.com/Azure/azure-sdk-for-go/services/network/mgmt/2021-08-01/network.LoadBalancerSkuName to func(sigs.k8s.io/cluster-api-provider-azure/api/v1beta1.SKU) github.com/Azure/azure-sdk-for-go/sdk/resourcemanager/network/armnetwork/v4.LoadBalancerSKUName
```

nojnhuh commented on azure/converters/loadbalancers.go:
```diff
@@ -17,14 +17,14 @@ limitations under the License.
 
-// SKUtoSDK converts infrav1.SKU into a network.LoadBalancerSkuName.
-func SKUtoSDK(src infrav1.SKU) network.LoadBalancerSkuName {
+// SKUtoSDK converts infrav1.SKU into an armnetwork.LoadBalancerSkuName.
```
nit:
```suggestion
// SKUtoSDK converts infrav1.SKU into an armnetwork.LoadBalancerSKUName.
```

nawazkh replied:
Updated.

nojnhuh commented on azure/services/loadbalancers/client.go:
```diff
@@ -18,158 +18,127 @@ package loadbalancers
+	auth          azure.Authorizer
 }
 
 // newClient creates a new load balancer client from subscription ID.
```
nit:
```suggestion
// newClient creates a new load balancer client from an authorizer.
```

nawazkh replied:
Updated

mboersma commented on azure/services/loadbalancers/client.go:
```diff
@@ -18,158 +18,127 @@ package loadbalancers
-	return loadBalancersClient
+	factory, err := armnetwork.NewClientFactory(auth.SubscriptionID(), auth.Token(), opts)
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to create load balancer client factory")
```
```suggestion
		return nil, errors.Wrap(err, "failed to create armnetwork client factory")
```

nawazkh replied:
Updated :)

mboersma reviewed (commented):
lgtm, I just had one change to a comment for consistency with the other services.

mboersma reviewed (approved):
/lgtm
/approve
